compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp

```diff
@@ -0,0 +1,645 @@
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Dialect/Utils/IndexingUtils.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#define DEBUG_TYPE "iree-opt-propagate-linalg-transpose"
+
+namespace mlir::iree_compiler::GlobalOptimization {
+
+//===----------------------------------------------------------------------===//
+// Transpose permutation helpers
+//===----------------------------------------------------------------------===//
+
+static bool isIdentityPermutation(ArrayRef<int64_t> perm) {
+  for (auto [index, dim] : llvm::enumerate(perm)) {
+    if (index != dim) {
+      return false;
+    }
+  }
+  return true;
+}
+
+// Constructs a transpose of the given tensor and permutation.
+static Value createTranspose(OpBuilder &builder, Value source,
+                             ArrayRef<int64_t> perm) {
+  SmallVector<OpFoldResult> mixedSizes =
+      tensor::getMixedSizes(builder, source.getLoc(), source);
+  applyPermutationToVector(mixedSizes, perm);
+  Type elemType = cast<RankedTensorType>(source.getType()).getElementType();
+  Value empty =
+      builder.create<tensor::EmptyOp>(source.getLoc(), mixedSizes, elemType)
+          .getResult();
+  return builder
+      .create<linalg::TransposeOp>(source.getLoc(), source, empty, perm)
+      ->getResult(0);
+}
+
+static RankedTensorType getPermutedTensorType(RankedTensorType type,
+                                              SmallVector<int64_t> perm) {
+  SmallVector<int64_t> permutedShape = applyPermutation(type.getShape(), perm);
+  return RankedTensorType::get(permutedShape, type.getElementType());
+}
+
+//===----------------------------------------------------------------------===//
+// Transpose specialization
+//===----------------------------------------------------------------------===//
+
+// Indicates whether the given linalg op represents a transpose. In particular,
+// it requires a single input where the indexing maps are full permutations and
+// non-equal.
+static bool isaTransposeOpInterface(linalg::LinalgOp linalgOp) {
+  // Structural.
+  if (linalgOp.getNumParallelLoops() != linalgOp.getNumLoops())
+    return false;
+
+  // Operands and maps.
+  if (linalgOp.getNumDpsInputs() != 1 || linalgOp.getNumDpsInits() != 1)
+    return false;
+  auto mapRange = linalgOp.getIndexingMapsArray();
+  if (mapRange.size() != 2 || !mapRange.front().isPermutation() ||
+      !mapRange.back().isPermutation() || mapRange.front() == mapRange.back()) {
+    return false;
+  }
+  // Region.
+  return llvm::hasSingleElement(linalgOp.getBlock()->getOperations());
+}
+
+// Checks whether the given linalg.generic is a transpose and specializes.
+static void specializeGenericTransposeOp(RewriterBase &rewriter,
+                                         linalg::GenericOp genericOp) {
+  if (isaTransposeOpInterface(genericOp)) {
+    auto mapRange = genericOp.getIndexingMapsArray();
```

Comment:
use early-return style, because it can save us a level of indents.

```suggestion
  if (!isaTransposeOpInterface(genericOp)) return;
  auto mapRange = genericOp.getIndexingMapsArray();
  ...
```